homeassistant/components/automation/__init__.py

```diff
@@ -228,6 +228,8 @@ def is_on(self) -> bool:
         if state:
             enable_automation = state.state == STATE_ON
             self._last_triggered = state.attributes.get('last_triggered')
+            if isinstance(self._last_triggered, str):
```

Comment:
It will always be a string?

Reply:
No, it might be `None`. This is the same code as is used here:

https://github.com/home-assistant/home-assistant/blob/1431fd6fbdd6aafd6fef0b41ea360ca44a7294bd/homeassistant/core.py#L751-L757

Reply:
If it can be None, can we then test using `is not None`.  That way we're more clear about what is happening.

Reply:
I would also prefer if we don't assign it to `self` until we have converted it. 

```python
last_triggered = state.attributes.get('last_triggered')
if last_triggered:
    self._last_triggered = parse_datetime(last_triggered)
```

Reply:
I guess that won't work. Apparently when automations are reloaded, the state that is retrieved in `state = await self.async_get_last_state()` has a `last_triggered` attribute that is already a `datetime`. When I made the change requested test `test_if_fires_on_hass_start`, which passed before, failed due to an exception:
```text
ERROR:homeassistant.core:Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/home/phil/repos/home-assistant/homeassistant/helpers/entity_platform.py", line 363, in _async_add_entity
    await entity.async_added_to_hass()
  File "/home/phil/repos/home-assistant/homeassistant/components/automation/__init__.py", line 232, in async_added_to_hass
    self._last_triggered = parse_datetime(last_triggered)
  File "/home/phil/repos/home-assistant/homeassistant/util/dt.py", line 124, in parse_datetime
    match = DATETIME_RE.match(dt_str)
TypeError: expected string or bytes-like object
```

Reply:
How about this:
```python
            try:
                self._last_triggered = parse_datetime(
                    state.attributes.get('last_triggered'))
            except TypeError:
                pass
```

Reply:
How about this:
```python
last_triggered = state.attributes.get('last_triggered')
try:
    self._last_triggered = parse_datetime(last_triggered)
except TypeError:
    self._last_triggered = last_triggered
```
But honestly, I think the way it was is clearer. If it's restoring a datetime that was JSON encoded, then it's simply doing the opposite -- i.e., it's effectively decoding the JSON. Maybe instead of changing the code add a comment to that effect???

Reply:
@balloob are you ok with my last suggestion?

Reply:
Let's not change the code to work around a failing test. Can you dig in why `last_triggered` would be a truthy value but it is not a string ? 

If the restored states don't have consistent typing, that's a bug.

Reply:
It's a valid test, and it's catching a valid problem (that was introduced using your suggestion.) Turns out the last_triggered attribute can be one of three types: None, a datetime, or a string (which is an ISO formatted datetime.) If async_added_to_hass is being called during restore, it can be None or a string. But after the automation.reload service is called, async_added_to_hass is called and the last_triggered attribute can be None or a datetime. So the only case where parse_datetime needs to be called is during a restore and if last_triggered is a string (and it will _only_ be a string during a restore.) So the original code, IMHO, makes the most sense, and adding a comment to explain what's happening is probably reasonable. And it's comparable to how this is handled elsewhere, as I originally explained.

Reply:
The bug is that `self.async_get_last_state()` is not consistent in what it returns. It should always be the same.

Reply:
I didn't know you'd want to go that deep, especially since there are a few places where parse_datetime is being called to compensate for the fact that a datetime is converted to an ISO formatted string when being saved, but the reverse is not done when it is restored.

I think the real bug is in homeassistant.helpers.storage.Store, where provisions are made for using an extended JSONEncoder when saving, but there are no provisions for using an extended JSONDecoder to properly decode everything. And in homeassistant.helpers.json where the extended JSONEncoder is defined, yet there's no extended JSONDecoder to match it. BTW, I don't think just encoding a datetime using its isoformat method is sufficient, because then attributes that were originally datetime's can't be distinguished from attributes that are purposely already a string created from datetime.isoformat().

Check out this article: https://www.techatbloomberg.com/blog/work-dates-time-python/, especially the section entitled "Serializing your datetime objects", and the sub-section "Objects".

Reply:
FWIW, I suspect any fix for the real issue (whichever it turns out to be) is going to be fairly involved. In the meantime, is it possible to address the problem mentioned in the description (which people have been working around for a while -- there have been several community forum discussions around this that I'm aware of) by accepting the original simple change in this PR, which works and solves the problem? The bigger issue can be then taken up separately?

Reply:
We're not going to add workarounds for bugs in our system, that is silly. 

https://github.com/home-assistant/home-assistant/pull/25073

Reply:
Added commit to assume it's always a string so it can go into the beta 👍